### ISSUE TITLE
Windows

### ISSUE THREAD

**Issue by nu8** [CakeML/cakeml#739]
If I compile like this:

~~~
$ make
cc    -c -o cake.o cake.S
cc    -c -o basis_ffi.o basis_ffi.c
cc  cake.o basis_ffi.o   -o cake.exe
~~~

I get this result in a MSYS2 console:

~~~
$ ./cake
sh: ./cake: cannot execute binary file: Exec format error
~~~

and this result in a Windows console:

~~~
cake.exe is not a valid Win32 application.
Access is denied.
~~~

This is my compiler:

<https://packages.msys2.org/base/mingw-w64-gcc>

**Comment by tanyongkiam** [comment]
Hi, I am guessing this is a problem with the wrapper script that I set up for Windows.

The first error in MSYS2 is a little concerning though. You are using a x64 machine and the 64-bit CakeML compiler right?

However, just as a quick thing to try, could you open up cake.S and change these lines:

`cake_heap:
     .space 1024 * 1024 * 1000
     .p2align 3
cake_stack:
     .space 1024 * 1024 * 1000
     .p2align 3`

To something like:

`cake_heap:
     .space 1024 * 1024 * 100
     .p2align 3
cake_stack:
     .space 1024 * 1024 * 100
     .p2align 3`

This has the effect of setting the default heap/stack size allocated to CakeML to 100MB instead of 1000MB.

**Comment by nu8** [comment]
That worked, but then this happened:

~~~
$ cat sunday.cml
print "Sunday";

$ make sunday.cake
cake.exe  <sunday.cml >sunday.cake.S
cc -E  sunday.cake.S > sunday.cake.s
as   -o sunday.cake.o sunday.cake.s
cc   sunday.cake.o   -o sunday.cake
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../
x86_64-w64-mingw32/lib/../lib/libmingw32.a(lib64_libmingw32_a-crt0_c.o): In
function `main': E:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/
crt/crt0_c.c:18: undefined reference to `WinMain'
~~~

I think because with Windows, `sunday.cake.S` and `sunday.cake.s` are the same
thing. So if I do this instead:

~~~
cake < sunday.cml > sunday.cake.S
cc -E sunday.cake.S > sunday.cake.T
as -o sunday.cake.o sunday.cake.T
cc sunday.cake.o -o sunday.cake
~~~

Get this result:

~~~
sunday.cake.o:cake.S:(.text+0x26562): undefined reference to `cml_exit'
sunday.cake.o:cake.S:(.text+0x26551): undefined reference to `ffiwrite'
~~~

**Comment by tanyongkiam** [comment]
Hmm that's a little weird, you'll need to link in basis_ffi.c (included in the zip) like what was doing for the compiler (in your first message).

Something like `cc foo.o basis_ffi.o`

I'm not sure why the Makefile doesn't seem to be doing that correctly here. Perhaps @xrchz would know.

**Comment by tanyongkiam** [comment]
BTW, the 1000->100 thing is turning up to be a common issue in that some compilers seem to accept the former fine but then the resulting executable has a runtime issue.

I have an ongoing PR that should somewhat fix this.

**Comment by nu8** [comment]
So if I edit `cake.S`, then do this:

~~~
cc -c cake.S
cc -c basis_ffi.c
cc -o cake cake.o basis_ffi.o
./cake < sunday.cml > sunday.S
cc -E sunday.S > sunday.T
as -o sunday.o sunday.T
cc sunday.o basis_ffi.o
~~~

The program compiles, but then it does the same thing Cake was doing:

~~~
cannot execute binary file: Exec format error
~~~

**Comment by tanyongkiam** [comment]
Ah, that's again because the default sunday.S produced by cake.S will again have 1000, 1000 for the heap and stack space.

It should work if you tweak sunday.S to also have 100, 100.

Alternatively, use `./cake --heap_size=100 --stack_size=100 < sunday.cml > sunday.S`

**Comment by nu8** [comment]
Ok I did this:

~~~
# EDIT cake.S
cc -c cake.S basis_ffi.c
cc -o cake cake.o basis_ffi.o
./cake < sunday.cml > sunday.S
# EDIT sunday.S
cc -E sunday.S > sunday.T
as -o sunday.o sunday.T
cc -o sunday sunday.o basis_ffi.o
./sunday
~~~

and it works. FYI it would go up to size `949` ok, higher than that it breaks.

**Comment by tanyongkiam** [comment]
Great! And thanks for investigating the size -- I think it's somehow dependent on the OS but I'm not too sure.

I don't understand why your system treats sunday.S and sunday.s as the same file, maybe it's case insensitive somehow.

**Comment by nu8** [comment]
Yeah, Windows doesnt pay attention to case. I think you can change that with Registry setting, but its not the default. So I would strongly suggest using `.asm` or some different extension for the generated file

**Comment by tanyongkiam** [comment]
Ah, I see. I think there's probably a workaround that can be made to the Makefile e.g., either not using .s for the intermediary file or using .asm instead of .S as you suggest.
